llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -314,9 +315,36 @@ bool RepoPruning::runOnModule(Module &M) {
     } while (IsEliminated);
   }
 
+  // Enable this pass support for alias.
+  for (Module::alias_iterator I = M.alias_begin(), E = M.alias_end(); I != E;) {
+    Module::alias_iterator CurI = I;
+    ++I;
+    auto *Aliasee = CurI->getBaseObject();
+    if (wasPruned(*Aliasee)) {
+      Type *Ty = CurI->getValueType();
+      // Delete the alias from Module if its aliasee is pruned.
+      CurI->removeFromParent();
+      // Replace the alias by a declaration.
```

Comment:
Is this always safe? Can LLVM only create aliases of values with external linkage? Should the replacement reflect the linkage of the original?

Reply:
> 
> 
> Is this always safe?
>
I think it is safe. Since an alias just a new symbol for an existing global value, a TicketNode with a `true` pruned is created for the alias during the RepoMetadataGeneration pass. Since it is always pruned, the alias will be written into the compilations. Therefore, it is safe to replace alias by a declaration. However, as you mentioned that the duplicated objects are generated. This need to be improved.
  
> 
> Can LLVM only create aliases of values with external linkage?
>
No, the alias may have any linkage types.

>
> Should the replacement reflect the linkage of the original?
>
No need. The original linkage type of an alias has been stored in its TicketNode, which will be used when generating the compilations. Here, the declaration with external linkage type will make it skip all the optimisition passes to reduce the compilation time.

Reply:
> Since an alias just a new symbol for an existing global value

Is it possible to have an aliases for a symbol with internal linkage?

Reply:
> Since an alias just a new symbol for an existing global value

Is it possible to have an aliases for a symbol with internal linkage?

Reply:
> 
> Is it possible to have an aliases for a symbol with internal linkage?
>
Yes, an alias could have the different linkage types which include the internal linkage. For example:
```  
@zed = global i32 42
@foo = internal alias i32, i32* @zed
@foo1 = weak alias i16, bitcast (i32* @zed to i16*)
```

Since alias must point to a definition, we can't keep the GO as alias if its aliasee has been pruned.
```
@zed2 = available_externally global i32 42
@zed3 = external global i32
@foo2 = internal alias i32, i32* @zed2
@foo3 = internal alias i32, i32* @zed3
```
Compile the IR:
```
llc -filetype=obj test.ll -o test.o
```
The following errors will be given:
```
Alias must point to a definition
i32* @foo2
Alias must point to a definition
i32* @foo3
C:\MyWork\git\github\llvm-project-prepo\build\Debug\bin\llc.exe: test.ll: error: input module is broken!
```

Reply:
> Since alias must point to a definition, we can't keep the GO as alias if its aliasee has been pruned.

This sounds like a problem. What are the implications?

Reply:
> 
> 
> > Since alias must point to a definition, we can't keep the GO as alias if its aliasee has been pruned.
> 
> This sounds like a problem. What are the implications?

I think it is fine with current patch. If an aliasee is pruned, it changes from definition to declaration. Its aliases will be changed to a declaration as well. Therefore, both aliasee and alias are declaration after pruning.